### ISSUE TITLE
/api/v1/read is not documented

### ISSUE THREAD

**Issue by shay-berman** [prometheus/prometheus#7192]
Two issues regarding remote_read feature:

1. Missing documentation about `/api/v1/read` in the api documentation page [here](https://prometheus.io/docs/prometheus/latest/querying/api/).

2. Also in the configuration page [here](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#configuration-file) 
  
```
# The labels to add to any time series or alerts when communicating with
# external systems (federation, remote storage, Alertmanager).
  external_labels:
    [ <labelname>: <labelvalue> ... ]
```

This does not give enough information that you cannot use `remote_read` with the conjunction  with  `external_labels` setting. Because if you do set in the remote_read prometheus a label but its different label then the prometheus instances that the remote_read prometheus is set - then the remote read does not work.

Consider to add some note in side the above link or maybe inside [remote_read](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read).

**Comment by brian-brazil** [comment]
> in the api documentation page here.

It's part of remote storage, which we haven't documented generally and wouldn't go on that page as it's not following our v1 API pattern.

> This does not give enough information that you cannot use remote_read with the conjunction with external_labels setting. 

You can use them together. This sounds more like a usage question.

**Comment by shay-berman** [comment]
Thanks @brian-brazil for the quick response.

Few questions please:
1. Why the team haven't documented the `remote_read` API?
2. Any impact if its not following v1 API pattern?
3. Is it recommended to get global view by having prometheus that just remote_read from many prometheus instances? (So Grafana will have single data source of all the targets collected by all the prometheus instances)

**Comment by roidelapluie** [comment]
In general the v1 documentation is under "querying" and it contains much more than querying (e.g. lifecycle). I guess some re-org could be done.

**Comment by brian-brazil** [comment]
Noone has sent a PR, and it's more a developer thing than a user thing.

The docs would be incorrect, as it's not JSON.

No, that would be a bad idea as you'll be limited by network bandwidth - and grafana can that that itself.

**Comment by shay-berman** [comment]
Thanks @brian-brazil for the clarification.

Today I have a big prometheus that scrap too many targets\metrics and a Grafana that use this prometheus as datasource for all the dashboards. My plan is to break it to multiple small prometheus instances(each scrap a range of targets), but then I want to have a global view on them.

**Which approach below is better?**
Approach1) to use prometheus that `remote_read` from all the small prometheus instances. And install the prometheus `remote_read` as k8s deployment with loadbalancer above to allow scale out depend on network traffic.

Approach2) to use Thanos [Querier](https://github.com/thanos-io/thanos/blob/master/docs/components/query.md#querier-use-cases-why-do-i-need-this-component) that read from all the small prometheus instances. And use this Thanos querier as datasource in Grafana.  I guess Thanos querier should have smaller footprint (vs prometheus `remote_read`) and its stateless by design so easier to scale it up to avoid network bandwidth limit.

**Comment by brian-brazil** [comment]
Usage questions like that are better suited to the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users).

**Comment by shay-berman** [comment]
Sure

**Comment by shay-berman** [comment]
@brian-brazil shell I close this issue or to keep it opened for improving the documentation of `/api/v1/read`?

**Comment by brian-brazil** [comment]
It should stay open, as this is indeed missing from the docs.

**Comment by hdost** [comment]
Is this something that should go under https://prometheus.io/docs/prometheus/latest/management_api/ or with this being added should it be re organized?

**Comment by hdost** [comment]
Also would something like https://github.com/swaggo/swag/blob/master/README.md be of interest?

**Comment by brian-brazil** [comment]
This probably requires a new section, though we might get away with just linking to the example adapter.

Swagger is a bigger discussion, and afaik also only covers JSON which this API is not.

**Comment by hdost** [comment]
👍🏼 ok on the Swagger part , and yes that project only supports JSON. The larger point was more about it might be best that the docs eventually get generated from sources in an effort to reduce drift. (Could be a separate issue)

As far as pointing to the example adapter I’m assuming this is what you’re meaning ‘documentation/examples/remote_storage/remote_storage_adapter/opentsdb/client.go’

I don’t think that would hurt to start, I think that would be a good supplement. 

I can work on a PR for a new doc page inspired by the management api page, but about remote read.

**Comment by hdost** [comment]
After an initial glance for now I think the simple format and reference by be the best method for now. I have started on a simple change, but I may make a different issue related to gRPC/REST api generation for documentation.

**Comment by hdost** [comment]
Actually it looks like there’s already an issue for the larger discussion in #2392

**Comment by brian-brazil** [comment]
Don't worry about the grpc stuff, that's deprecated and going away - there's no need to document it.

**Comment by hdost** [comment]
So I’ll just worry about the REST protobuf

**Comment by xiaozongyang** [comment]
About `/api/v1/read`, is there some option to reduce its latency? May it be the option `--storage.remote.read-sample-limit`, `--storage.remote.read-max-bytes-in-frame` and `--storage.remote.read-concurrent-limit`?

By the way, are there any existing issues talking about reducing remote read latency?

**Comment by brian-brazil** [comment]
Please don't ask support questions on unrelated issues, the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users) is the place for those.

**Comment by xiaozongyang** [comment]
Got it, sorry about that. I will ask this question on mailing list.

**Comment by chenlujjj** [comment]
Any update on this issue ?

**Comment by bzed** [comment]
Indeed, a documentation (or even better: some easy to include code) of the remote read API would be very appreciated.
For example there are issues like https://github.com/go-graphite/graphite-clickhouse/issues/99 which are basically struggling with verifying if the implementation after the upgrade works as expected. And so far I was neither able to find a documentation of what is expected from the API, nor is there some regression test suite one can run against an external API. carbon-clickhouse/graphite-clickhouse works very well as prometheus remote storage, but keeping it uptodate and being able to verify/CI changes would be necessary.

**Comment by skupjoe** [comment]
I spent days troubleshooting why remote read was not working (in this case, M3DB -> Promethesu), and it turned out to be because I was using `external_labels` in my global config.  As described in the OP, this just flat-out breaks remote read from working (but Prometheus will still accept remote writes no problem).

I wish this was documented better- it would have saved me a lot of time.

Also not mentioned is the fact that metrics browser (dropdown) in Prometheus/Grafana don't display any remote read timeseries- whether it is working or not. This added additional confusion.

**Comment by TaylorMutch** [comment]
From @skupjoe 
> I spent days troubleshooting why remote read was not working (in this case, M3DB -> Promethesu), and it turned out to be because I was using external_labels in my global config. As described in the OP, this just flat-out breaks remote read from working (but Prometheus will still accept remote writes no problem) and there are no errors displayed in logs, nothing.

I just went through the same thing using https://github.com/jacksontj/promxy and had `external_labels` set in my global config; removing that configuration made everything work.

**Comment by krajorama** [comment]
Hello from the bug scrub: the [remote read API documentation](https://prometheus.io/docs/prometheus/latest/querying/remote_read_api/#remote-read-api-0) is in place, however keeping open as still relevant for the external labels semantics documentation. In the bug scrub we spent a short time reading documentation but couldn't figure out the exact behavior between using external labels and remote read (nor Prometheus exposing remote read neither using remote read). We suggest the following acceptance criteria :  add documentation for behavior on external_labels with remote read exposing and using.

**Comment by Shivampal157** [comment]
Hi maintainers,

I’m planning to contribute a focused documentation update for /api/v1/read
covering its behavior, limitations, and common pitfalls (especially around
external_labels and visibility in Prometheus/Grafana), rather than a full API spec.

I’ll start with a small, reviewable docs PR if that sounds reasonable.
Thanks!
